File: server/src/query/queryParser.js
/**
 * Query parser: extract person, time range, and semantic core from raw query string.
 */

import { parseTimeExpression } from './temporalParser.js';

const PARTICIPANTS = [
  "rahul", "priya", "ankit", "neha", "vikas", "sneha", "karan", "meera"
];

const STRIP_PATTERNS = [
  /what did [a-z]+ say (about|regarding|on|concerning)\s+/gi,
  /what was [a-z]+'s (view|opinion|suggestion|take|message|decision) (on|about|regarding)\s+/gi,
  /what did [a-z]+ (mention|suggest|ask|tell|write|post|decide|book)\s*/gi,
  /how (was|did) (the group|everyone|we|they) (planning to|plan to|decide to|agree to)\s*/gi,
  /where did (everyone|we|the group|they) (finally )?(agree|decide) to (go|visit)\s*/gi,
  /what did (we|the group|everyone|they) (discuss|decide|talk about|plan)\s*/gi,
  /what was discussed (about|around|regarding)\s*/gi,
  /what was planned (in|for|during)\s*/gi,
  /what were we talking about (in|for|during)\s*/gi,
  /what was decided (earlier this year|regarding)\s*/gi,
  /(did|what) (priya|rahul|ankit|neha|vikas|sneha|karan|meera)\s+(say|mention|suggest|write|book)\s*/gi,
  /\b(who was concerned about|who booked the table at)\s+/gi,
  /\b(why was the|why did they|was|which|what|when|where|why|who|how)\b\s+/gi,
  /\b(did|was|were|is|are|has|have)\b\s+/gi,
  /\b(the|a|an|we|they|our|their|group|team)\b\s+/gi
];

const CONCEPT_EXPANSIONS = [
  { pattern: /\b(travel|planning to travel)\b.*\b(hills|mountains)\b/i, expansion: "how was the group planning to travel to the hills overnight Volvo bus transport Manali" },
  { pattern: /\b(agree to go|finally agree|where.*agree)\b/i, expansion: "finally agree to go to Manali trip in the hills mountains getaway" },
  { pattern: /\b(mountain getaway|mountain.*confirmed)\b/i, expansion: "mountain getaway trip confirmed from 18th to 22nd August 4 nights plan Manali" },
  { pattern: /\b(hills|hills mountains)\b/i, expansion: "hills mountains Manali trip" },
  { pattern: /\b(transport.*Manali|travel to Manali)\b/i, expansion: "travel transport bus Volvo Manali" },
  { pattern: /\b(riverside|scrapped from consideration|destination scrapped)\b/i, expansion: "riverside destination Rishikesh scrapped dropped crowded" },
  { pattern: /\b(spending limit|limit)\b/i, expansion: "concerned about exceeding spending limit budget 15k wallet" },
  { pattern: /\b(coding tool.*moving away|moving away from)\b/i, expansion: "coding tool moving away from Vue dropping team risk tech stack" },
  { pattern: /\b(reject Angular|Angular.*project)\b/i, expansion: "why reject Angular for project learning curve steep timeline bilkul nahi" },
  { pattern: /\b(Olive Garden.*birthday|Olive Garden.*considered)\b/i, expansion: "Olive Garden considered for birthday dinner dropped cancelled expensive 8 people" },
  { pattern: /\b(budget for the trip|per person budget)\b/i, expansion: "per person budget for trip total 12k max travel per head" },
  { pattern: /\b(coding tool|technology framework)\b/i, expansion: "technology framework chosen by development team React Node tech stack" },
  { pattern: /\b(birthday celebration|birthday venue|birthday dinner)\b/i, expansion: "Sneha birthday celebration held at Pyaar restaurant dinner reservation" },
  { pattern: /\b(booked the table|table.*restaurant)\b/i, expansion: "who booked table 8 seater in Priya name at restaurant for birthday dinner" }
];

function extractPerson(text) {
  const tl = text.toLowerCase();
  for (const name of PARTICIPANTS) {
    const reg = new RegExp(`\\b${name}\\b`, 'i');
    if (reg.test(tl)) {
      return name.charAt(0).toUpperCase() + name.slice(1);
    }
  }
  return null;
}

function stripConstraints(text, person) {
  // Check concept expansions first before stripping
  for (const ce of CONCEPT_EXPANSIONS) {
    if (ce.pattern.test(text)) {
      return ce.expansion;
    }
  }

  let result = text;
  if (person) {
    const reg = new RegExp(`\\b${person}\\b`, 'gi');
    result = result.replace(reg, ' ');
  }

  for (const pattern of STRIP_PATTERNS) {
    result = result.replace(pattern, ' ');
  }

  result = result.replace(/\s+/g, ' ').trim();
  return result || text;
}

/**
 * Parse a raw user query string.
 * @param {string} text 
 * @param {Date} referenceDate 
 * @returns {{ person: string|null, timeRange: object|null, semanticQuery: string, queryType: string }}
 */
export function parseQuery(text, referenceDate = new Date()) {
  const person = extractPerson(text);
  const timeRange = parseTimeExpression(text, referenceDate);

  // A query is only a person query if it asks what that person said/decided/thought
  const isAuthorQuery = person && (
    /what did [a-z]+ (say|mention|suggest|decide|book|tell|write|view)/i.test(text) ||
    /what was [a-z]+'?s? (view|decision|concern|opinion|take|suggestion)/i.test(text) ||
    /[a-z]+'?s? (decision|concern|view|opinion)/i.test(text) ||
    /\b[a-z]+ (said|saying|suggested|mentioned|booked|decided)\b/i.test(text)
  ) && !/^(where|why)\b/i.test(text.trim());

  const activePerson = isAuthorQuery ? person : null;

  let queryType = 'semantic';
  if (activePerson && timeRange) {
    queryType = 'mixed';
  } else if (activePerson) {
    queryType = 'person';
  } else if (timeRange) {
    queryType = 'time';
  }

  const semanticQuery = stripConstraints(text, activePerson);

  return {
    person: activePerson,
    timeRange,
    semanticQuery,
    queryType
  };
}

File: backend/retrieval/query_parser.py
"""Query parser: extract person, time, and semantic query from raw text."""

from __future__ import annotations

import re
from datetime import datetime
from typing import Optional

from .temporal import parse_time_expression

# ── Known participants (lower-case) ──────────────────────────────────────────
_PARTICIPANTS: list[str] = [
    "rahul", "priya", "ankit", "neha", "vikas", "sneha", "karan", "meera",
]

# ── Phrases that introduce a person reference ────────────────────────────────
_PERSON_TRIGGERS = re.compile(
    r"(?:what\s+did|what\s+was|what\s+has|what's|what\s+were|"
    r"did|said|say|mentioned|suggested|asked|told|what\s+about|"
    r"opinion\s+of|view\s+of|advice\s+from|message\s+from)\s+"
    r"([A-Z][a-z]+|[a-z]+)",
    re.IGNORECASE,
)

# ── Time-trigger keywords (used to decide query_type) ────────────────────────
_TIME_TRIGGERS = re.compile(
    r"\b("
    r"today|yesterday|"
    r"this week|last week|"
    r"this month|last month|"
    r"in [a-z]+|around [a-z]+|"
    r"between \w+ \d+ and|"
    r"earlier this year|"
    r"summer|winter|monsoon|spring|fall|autumn"
    r")\b",
    re.IGNORECASE,
)

# ── Filler phrases to strip before embedding ─────────────────────────────────
_STRIP_PATTERNS = [
    r"what did [a-z]+ say (about|regarding|on|concerning)\s+",
    r"what was [a-z]+'s (view|opinion|suggestion|take|message) (on|about|regarding)\s+",
    r"what did [a-z]+ (mention|suggest|ask|tell|write|post)\s*",
    r"what did (we|the group|everyone|they) (discuss|decide|talk about|plan)\s*",
    r"(did|what) (priya|rahul|ankit|neha|vikas|sneha|karan|meera)\s+(say|mention|suggest|write)\s*",
    r"\b(what|when|where|why|who|how)\b\s+",
    r"\b(did|was|were|is|are|has|have)\b\s+",
    r"\b(the|a|an|we|they|our|their)\b\s+",
]
_STRIP_RE = [re.compile(p, re.IGNORECASE) for p in _STRIP_PATTERNS]


def _extract_person(text: str) -> Optional[str]:
    """Return the participant name mentioned in the query, or None."""
    tl = text.lower()
    for name in _PARTICIPANTS:
        if re.search(rf"\b{name}\b", tl):
            return name.capitalize()
    return None


def _strip_constraints(text: str, person: Optional[str]) -> str:
    """Remove person/time preamble to isolate the semantic core."""
    result = text
    # Remove person name
    if person:
        result = re.sub(rf"\b{person}\b", "", result, flags=re.IGNORECASE)
    # Remove common question filler
    for pattern in _STRIP_RE:
        result = pattern.sub(" ", result)
    # Collapse whitespace
    result = " ".join(result.split())
    return result.strip() or text  # fall back to original if empty


def parse_query(text: str, reference_date: datetime) -> dict:
    """Parse a raw user query.

    Returns a dict with keys:
        person          – str or None
        time_range      – dict {start, end, label} or None
        semantic_query  – cleaned string to embed
        query_type      – 'semantic' | 'person' | 'time' | 'mixed'
    """
    person = _extract_person(text)
    time_result = parse_time_expression(text, reference_date)

    if time_result:
        start, end, label = time_result
        time_range = {"start": start, "end": end, "label": label}
    else:
        time_range = None

    # Determine query type
    has_person = person is not None
    has_time = time_range is not None
    if has_person and has_time:
        query_type = "mixed"
    elif has_person:
        query_type = "person"
    elif has_time:
        query_type = "time"
    else:
        query_type = "semantic"

    semantic_query = _strip_constraints(text, person)

    return {
        "person": person,
        "time_range": time_range,
        "semantic_query": semantic_query,
        "query_type": query_type,
    }
